test/MultiRole.js

```diff
@@ -0,0 +1,166 @@
+    assert.isTrue(await multiRole.holdsRole("1", account3));
+    await multiRole.revertIfNotHoldingRole("1", { from: account3 });
+
+    // Anyone who holds the role should be able to remove members.
+    await multiRole.removeMember("1", account2, { from: account3 });
+    assert.isFalse(await multiRole.holdsRole("1", account2));
+    assert(await didContractThrow(multiRole.revertIfNotHoldingRole("1", { from: account2 })));
+  });
+
+  it("Shared externally-managed (shared) role", async function() {
+    const multiRole = await MultiRoleTest.new();
+    await multiRole.createSharedRole("1", "1", [account1, account3]);
+
+    // Roles can only be managed by previously created roles.
+    assert(await didContractThrow(multiRole.createSharedRole("2", "3", [account1, account2])));
+    await multiRole.createSharedRole("2", "1", [account1, account2]);
+
+    // Check that only the account1 and account2 are holders of the role.
+    assert.isTrue(await multiRole.holdsRole("2", account1));
+    assert.isTrue(await multiRole.holdsRole("2", account2));
+    assert.isFalse(await multiRole.holdsRole("2", account3));
+    await multiRole.revertIfNotHoldingRole("2", { from: account1 });
+    await multiRole.revertIfNotHoldingRole("2", { from: account2 });
+    assert(await didContractThrow(multiRole.revertIfNotHoldingRole("2", { from: account3 })));
+
+    // Anyone who holds the managing role should be able to add members.
+    assert(await didContractThrow(multiRole.addMember("2", account3, { from: account2 })));
+    await multiRole.addMember("2", account3, { from: account3 });
+    assert.isTrue(await multiRole.holdsRole("2", account3));
+    await multiRole.revertIfNotHoldingRole("2", { from: account3 });
+
+    // Anyone who holds the managing role should be able to remove members.
+    assert(await didContractThrow(multiRole.removeMember("2", account2, { from: account2 })));
+    await multiRole.removeMember("2", account2, { from: account1 });
+    assert.isFalse(await multiRole.holdsRole("2", account2));
+    assert(await didContractThrow(multiRole.revertIfNotHoldingRole("2", { from: account2 })));
+  });
+
+  it("Shared externally-managed (exclusive) role", async function() {
+    const multiRole = await MultiRoleTest.new();
+    await multiRole.createExclusiveRole("1", "1", account1);
+    await multiRole.createSharedRole("2", "1", [account2, account3]);
+
+    // Check that only the account2 and account3 are holders of the role.
+    assert.isTrue(await multiRole.holdsRole("2", account2));
+    assert.isTrue(await multiRole.holdsRole("2", account3));
+    assert.isFalse(await multiRole.holdsRole("2", account1));
+    await multiRole.revertIfNotHoldingRole("2", { from: account2 });
+    await multiRole.revertIfNotHoldingRole("2", { from: account3 });
+    assert(await didContractThrow(multiRole.revertIfNotHoldingRole("2", { from: account1 })));
+
+    // Anyone who holds the managing role should be able to add members.
```

Comment:
ultra nit: maybe reword this comment to "Only the holder of the managing role...", because the managing role is an exclusive role.

Reply:
Done.